### ISSUE TITLE
P2P Name System

### ISSUE THREAD

**Issue by jerrygreen** [hypercore-protocol/hypercore#283]
I would like to provide some domain, say:

https://example.com/

And I would like users to set some slugs names, for example:

https://example.com/jerrygreen
https://example.com/andrewosh
https://example.com/pfrazee
etc.

Each of these slug urls is a page associated with a hyperdrive, with some data related to a particular user:

https://example.com/jerrygreen => hyper://dffaf9446381f74a2406b22a9d56d59a31aed27d99c72c34a70b2ab49c7c39c6
https://example.com/andrewosh => hyper://1ae995fce0fb1ff8881466845187e110aadfe33a038bf807a1e76ff9636707c9
https://example.com/pfrazee => hyper://684bc636194748d6f9d637cae2790b4773ad410c32468301654ee8110827adf2
etc.

Of course I know how I can achieve this in traditional way: setup a server, a key-value database (or whatever else type of database, actually), set an UNIQUE constraint, let users write to this database, and use this database for routing. Easy. But it's technically my data, and it becomes source of truth, coming from my server: it contradicts being fully P2P.

Instead, I want this hash to be determined by users, and somehow distributed among users, so my server isn't a source of truth but, like... user of truth 😅 I see the problem is related to merkle trees, i.e. hugely into what hypercore team does already, so I'm here to ask how to properly approach this problem (P2P DNS / P2P user slugs), because even though I'm in web development my whole life, I'm new to developing in P2P environment, and I don't see existing solutions to that.

**Comment by aral** [comment]
Would recommend using subdomains and registering your main domain on the public suffix list. Then you become a de facto TLD. This is the Small Web approach. Will be releasing a more formal definition/protocol soon but in the meanwhile this might give you some ideas:

https://ar.al/2020/08/07/what-is-the-small-web/

**Comment by derhuerst** [comment]
@JerryGreen

To have the P2P-based data as a source of truth, you could let people choose their own slug/handle/subdomain  (e.g. `jerrygreen.awesome-p2p.app` or `awesome-p2p.app/jerrygreen` or `awesome-p2p-app://jerrygreen`) and tie it to a P2P-based data source *they* control cryptographically, e.g. *their* `hyper://…` feed. This prevents you from acting against their interest, because you can't tie unwanted slugs/handles/subdomains to users, as they'd have to acknowledge that explicitly. This mechanism could be implemented by **letting users publish a `awesome-p2p.json` "manifest" in their `hyper://…` feed**.

You should also prevent abuse in the other direction: People could arbitrarily claim slugs/handles/subdomains that someone else has already (legitimately) claimed, or should be "banned"/unlisted by removing/invalidating their slug/handle/subdomain. You would have to establish a **trusted source of "acknowledged"/listed slugs/handles/subdomains and their respective `hyper://…` feeds**. Technically, there are many ways to do this, e.g. by using [dat `TXT` DNS records](https://beakerbrowser.com/docs/guides/use-a-domain-name-with-dat#dat-dns-txt-records), or [serving `.well-known/dat` via `https://`](https://beakerbrowser.com/docs/guides/use-a-domain-name-with-dat#well-knowndat), or by adding one JSON file per "acknowledged"/listed slug/handle/subdomain to an authoritative `hyper://…` feed.

To enable accessibility via both Dat-based tools *and* `https://`, you can use the common [mechanisms to tie a (sub)domain or `https://` resource to a Dat feed](https://beakerbrowser.com/docs/guides/use-a-domain-name-with-dat#well-knowndat).

**Comment by csuwildcat** [comment]
> Would recommend using subdomains and registering your main domain on the public suffix list. Then you become a de facto TLD. This is the Small Web approach. Will be releasing a more formal definition/protocol soon but in the meanwhile this might give you some ideas:
> 
> https://ar.al/2020/08/07/what-is-the-small-web/

What is the assumption around global (planet-wide) resolution of peer IDs across all nodes that exist any any given time? I see (what looks like) a traditional DNS domain being suggested as a basis for orientation/resolution, so what is the suggested relationship between traditional DNS domains and IDs in the peer network? Finally, what guarantees are made about peer ID resolution across all possible nodes - can I review any proofs (formal or otherwise) that there is no (at least in the realm of virtual probabilistic certainty) need to worry about segmentation, state withholding, collusive node practices, or other issues that may affect key sets and other PKI metadata when IDs are resolved?

**Comment by jerrygreen** [comment]
I edited my initial post: I removed the [confusing] introduction, and also renamed the topic:

<s>P2P Domain Name System</p> -> P2P Name System

To get rid of potential confusion that the problem is closely related to existing DNS somehow. It’s loosely related, because what I want:

- Make the name system under my app to be P2P rather than single source database. This kinda _mimics_ the existing DNS in a way that existing DNS for web is already P2P, but no more related than that.

The rest part was confusing so I deleted this, and it was probably a separate issue [i.e. not related].
